### PR TITLE
Resolve IMMUTABLE crash on android 12 following a change in android specification

### PR DESCRIPTION
**Pull request by Octoton**
Now that the app is targeting api 31, pendingIntent need to be flag with FLAG_MUTTABLE or FLAG_IMMUTABLE or the app will crash: https://developer.android.com/guide/components/intents-filters#DeclareMutabilityPendingIntent

### CODE REVIEW

**Comment by soncaokim** [comment]
Can these changes wrapped into a PendingIntentCompat helper class please?

**Comment by Octoton** [comment]
Even if this is not always the same static method that is called ? Moreover, Flag_mutable could be use sometime (not very likely tho)

**Comment by soncaokim** [comment]
> Even if this is not always the same static method that is called ? Moreover, Flag_mutable could be use sometime (not very likely tho)

Something like https://github.com/M66B/FairEmail/blob/master/app/src/main/java/eu/faircode/email/PendingIntentCompat.java

It's your call: I'm okay to stick with the current proposal since it's clear enough.

**Review by soncaokim — APPROVED** [review]

File: app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl24.java [hunk 1]
```diff
@@ -50,12 +51,12 @@ public synchronized void update() {
...
         Intent action = new Intent(service, MainActivity.class);
         action.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
-        final PendingIntent clickIntent = PendingIntent.getActivity(service, 0, action, 0);
+        final PendingIntent clickIntent = PendingIntentCompat.getActivity(service, 0, action, 0);
 
         final ComponentName serviceName = new ComponentName(service, MusicService.class);
         Intent intent = new Intent(MusicService.ACTION_QUIT);
         intent.setComponent(serviceName);
-        final PendingIntent deleteIntent = PendingIntent.getService(service, 0, intent, 0);
+        final PendingIntent deleteIntent = PendingIntentCompat.getService(service, 0, action, 0);
```
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl24.java:59]: Is the change on 3rd parameter on purpose? ie.  from `intent` to `action`

**Comment by Octoton** [comment]
@soncaokim here is the clean fix (sory for the delayed response, wasn't home those few day), this PR should resolve android 12 crash and thus we shouldn't have to merge https://github.com/AdrienPoupa/VinylMusicPlayer/pull/552

File: app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl24.java (hunk 1 above)
Comment (Octoton) [reply app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl24.java:59]: Good catch

Comment (Octoton) [reply app/src/main/java/com/poupa/vinylmusicplayer/service/notification/PlayingNotificationImpl24.java]: Corrected

**Comment by soncaokim** [comment]
> @soncaokim here is the clean fix (sory for the delayed response, wasn't home those few day), this PR should resolve android 12 crash and thus we shouldn't have to merge #552

Thanks for the reaction. 

However I'm unsure if this is related to  the song deletion regression. I'm having the trouble on a device running Android 11, and it's not crashing, just does not delete the requested song from the device.

The regression actually doesnt happen right away on installing a new APK, but after sometime. 
The folder access permission dialog pops up, acts if it understood my choice to grant access to a folder, but fails silently afterwards.

A rollback to API 29 solves it.

**Comment by Octoton** [comment]
> > @soncaokim here is the clean fix (sory for the delayed response, wasn't home those few day), this PR should resolve android 12 crash and thus we shouldn't have to merge #552
> 
> Thanks for the reaction.
> 
> However I'm unsure if this is related to the song deletion regression. I'm having the trouble on a device running Android 11, and it's not crashing, just does not delete the requested song from the device.
> 
> The regression actually doesnt happen right away on installing a new APK, but after sometime. The folder access permission dialog pops up, acts if it understood my choice to grant access to a folder, but fails silently afterwards.
> 
> A rollback to API 29 solves it.

Ok my bad, I red the title to fast and didn't understood this was related to something else (could you add more info into this PR then ?)

**Comment by Octoton** [comment]
@AdrienPoupa can this be merge as without I cannot even open the master branch on android 12

**Comment by soncaokim** [comment]
> @AdrienPoupa can this be merge as without I cannot even open the master branch on android 12

About targeting Api 31 / Android 12: are you able to delete tracks and to add songs to playlists/faves on Android 12? To me there is a regression with these features, and I need to rollback to target Api 29.

**Comment by Octoton** [comment]
> > @AdrienPoupa can this be merge as without I cannot even open the master branch on android 12
> 
> About targeting Api 31 / Android 12: are you able to delete tracks and to add songs to playlists/faves on Android 12? To me there is a regression with these features, and I need to rollback to target Api 29.

Me neither, we clearly need to rollback to api 29, until we have corrected this problem (or the foreground crash in android 12). 
However, this PR will be necessary when we decide to target android 12 in the futur, thus it stand to reason that we merge it now as it doesn't impact older api.

**Comment by Octoton** [comment]
Moreover, I cannot change a song information with api 31 (change title or picture, stuff like that). Instead the app show me an old activity:
Screenshot_20220106-100233

**Comment by soncaokim** [comment]
> Me neither, we clearly need to rollback to api 29, until we have corrected this problem (or the foreground crash in android 12).
> However, this PR will be necessary when we decide to target android 12 in the futur, thus it stand to reason that we merge it now as it doesn't impact older api.

I think there is some minor change for this PR so that it can be built on API 29 - see https://app.circleci.com/pipelines/github/AdrienPoupa/VinylMusicPlayer/787/workflows/d2963848-6082-4f9a-ad08-14defa971004/jobs/930?invite=true#step-106-49

That's a CI build where I've applied this PR along the rollback of API 31 to 29

**Comment by Octoton** [comment]
Will check on this

**Comment by Octoton** [comment]
Added a commit for rollback to api 29, not the prettiest as it lessen the flexibility of this new class but FLAG_MUTABLE was introduce in api 31 so there is not really any other way

**Comment by soncaokim** [comment]
Merging this since it's a crash fix

**Pull request merged**